interfaces/attrs.go

```diff
@@ -0,0 +1,197 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+type PlugData struct {
+	plug         *snap.PlugInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type SlotData struct {
+	slot         *snap.SlotInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type Attributes interface {
+	SetStaticAttr(key string, value interface{})
+	StaticAttr(key string) (interface{}, error)
+	StaticAttrs() map[string]interface{}
+	Attr(key string) (interface{}, error)
+	Attrs() (map[string]interface{}, error)
+	SetAttr(key string, value interface{}) error
+}
+
+func NewSlotData(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) *SlotData {
+	return &SlotData{
+		slot:         slot,
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+func NewPlugData(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) *PlugData {
+	return &PlugData{
+		plug:         plug,
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+func (attrs *PlugData) Interface() string {
+	return attrs.plug.Interface
+}
+
+func (attrs *PlugData) Name() string {
+	return attrs.plug.Name
+}
+
+func (attrs *PlugData) Ref() PlugRef {
+	return PlugRef{Snap: attrs.plug.Snap.Name(), Name: attrs.plug.Name}
```

Comment:
That's just `attrs.plug.Ref()`

Reply:
No.. For some reason we have Ref() on Plug and Slot, but not on PlugInfo/SlotInfo. Perhaps we could have that on both for convienience, in which case this can be simplified indeed?

Reply:
Yes, please move them layer below.